### ISSUE TITLE
Same schema registered to multiple subjects - multiple global IDs - different behavior compared to Confluent Schema Registry

### ISSUE THREAD

**Issue by forsberg** [Apicurio/apicurio-registry#796]
Consider the following example schema:

```
{"schema": "{\"type\": \"record\", \"name\": \"ID\", \"namespace\": \"com.example\", \"fields\": [{\"doc\": \"ID\", \"name\": \"id\", \"type\": \"int\"}]}"}
```

If I register this with Confluent Schema Registry, on subject "test", I get the following result:

```
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8081/subjects/test/versions' -H 'Content-Type: application/vnd.schemaregistry.v1+json' -d@id-schema-namespaced.json ; echo
{"id":1}
```

Now, if I re-register it but on a different subject "test2":

```
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8081/subjects/test2/versions' -H 'Content-Type: application/vnd.schemaregistry.v1+json' -d@id-schema-namespaced.json ; echo
{"id":1}
```

We can see that the schema is now registered on two different subjects:

```
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8081/subjects/test/versions' ; echo
[1]
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8081/subjects/test/versions/1' ; echo
{"subject":"test","version":1,"id":1,"schema":"{\"type\":\"record\",\"name\":\"ID\",\"namespace\":\"com.example\",\"fields\":[{\"name\":\"id\",\"type\":\"int\",\"doc\":\"ID\"}]}"}
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8081/subjects/test2/versions' ; echo
[1]
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8081/subjects/test2/versions/1' ; echo
{"subject":"test2","version":1,"id":1,"schema":"{\"type\":\"record\",\"name\":\"ID\",\"namespace\":\"com.example\",\"fields\":[{\"name\":\"id\",\"type\":\"int\",\"doc\":\"ID\"}]}"}
```

As we can see - even though the schema is registered on two different subjects, it has the same **Global ID**: 1

However, Apicurio Registry 1.3.0.Final, and I've tested this with both the asyncmem and the streams backend, behaves differently:

```
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8080/api/ccompat/subjects/test/versions' -H 'Content-Type: application/vnd.schemaregistry.v1+json' -d@id-schema-namespaced.json ; echo
{"id":1}
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8080/api/ccompat/subjects/test2/versions' -H 'Content-Type: application/vnd.schemaregistry.v1+json' -d@id-schema-namespaced.json ; echo
{"id":2}
```

We can see already that the global ID is not the same, and this can be verified by using the versions endpoint:

```
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8080/api/ccompat/subjects/test/versions/1' ; echo
{"id":1,"subject":"test","version":1,"schema":"{\"type\": \"record\", \"name\": \"ID\", \"namespace\": \"com.example\", \"fields\": [{\"doc\": \"ID\", \"name\": \"id\", \"type\": \"int\"}]}"}
$ curl  -H 'Accept: application/vnd.schemaregistry.v1+json' 'http://localhost:8080/api/ccompat/subjects/test2/versions/1' ; echo
{"id":2,"subject":"test2","version":1,"schema":"{\"type\": \"record\", \"name\": \"ID\", \"namespace\": \"com.example\", \"fields\": [{\"doc\": \"ID\", \"name\": \"id\", \"type\": \"int\"}]}"}
```

Is this a bug, or by design? If it's by design, why? 

For me, having an ID that is the same for a specific schema, regardless of subject is part of the beauty of using a Schema Registry, as it gives me flexibility in how I use schemas and route data.

**Comment by EricWittmann** [comment]
The behavior you're seeing is "expected" but I wouldn't go so far as to say it was "as designed" - I think that implies that it's a deliberate feature.  It's more of a result of our implementation.  In our view, the global ID of any particular schema version is immutable but not guaranteed to be the same if you upload the same schema multiple times as different subjects.  It's true that this is a deviation from the Confluent Schema Registry implementation.

If you upload the same content version within a single Artifact, our API gives you the option to choose how that should be handled.  See the "ifExists" parameter here:  https://studio-ws.apicur.io/sharing/76f236d3-1312-4548-a4d1-400487188c66#operation/createArtifact

I always saw the Global ID as more of an implementation detail.  There are certainly some advantages to ensuring you only store identical content once rather than repeatedly.  Although the benefits of that would only really matter if you used the same schema across **a lot** of subjects.

Can you explain this more:

> For me, having an ID that is the same for a specific schema, regardless of subject is part of the beauty of using a Schema Registry, as it gives me flexibility in how I use schemas and route data.

I don't see why what I think is a storage optimization would have this level of impact.  What flexibility does it give you?

**Comment by EricWittmann** [comment]
I would also be interested in thoughts from @Apicurio/developers and @Apicurio/ibm-developers !

**Comment by forsberg** [comment]
The use case where I see global schema IDs that stay the same regardless of how many topics they are added to is when the data is generated by IoT devices that don't have access to the schema registry after build time. I've been part of projects where we've built Android apps, and embedded devices  reporting data to a central data collection cloud as Avro datums with the binary serialization, adding a header that contains only the schema ID.

In these cases, the schema has been used at build-time to generate a class in the target language, including a constant that defines which schema ID to add as a header when reporting data, and that's all that's needed - neither copy of schema (.avsc file) nor any contact with schema registry has been necessary, post-build.

The typical scenario is the code for the app/embedded device is compiled once, and that's where we add the schema ID, then it's put into a staging environment where it reports to a staging cloud which routes it into a staging Kafka topic. Once that has been verified, configuration is modified to have it report to the production cloud, which will then route it into the proper Kafka topic (the schema will have to be added to the corresponding schema subject, but that can easily be done as part of CI/CD).

Having to inject a new schema ID *in the client* at this stage is cumbersome and error-prone, I much prefer to have the schema ID kept inside my client artifact together with the compiled classes, where it simply can't be modified. That's why I prefer "One schema, one global ID"

**Comment by EricWittmann** [comment]
I see - then I assume the data/schemas are separated from the apps such that the data layer has its own build?  For example into a library that is then pulled into the apps when necessary?  And then presumably the data is published to a per-app topic rather than a per-schema topic.

Because if the build is happening per-app then everything would hang together just fine using Apicurio Registry - the app build would be responsible for publishing the schema to the registry in the way that you describe.  The only difference would be that multiple schemas (with the same content) would be published and therefore multiple global IDs would be created and packaged with the apps.

This seems a little logically frail.  I would need to read about the guarantees that the Confluent Schema Registry offers around global ID, but for this use case it seems like you'd be much better off using a crypto hash of the schema rather than a globalID.  Basically canonicalize the schema content and generate e.g. a SHA hash from that.  The registry does the same thing, indexing the content by that hash.  That way it wouldn't matter which Registry the schema was registered with, the content hash would always be the same (assuming the same canonicalizer).

For global ID I'm worried about this moving of schema content from environment to environment and continuing to rely on that ID.  In our system the ID is generated when the content is stored, based on the storage being used (e.g. an auto-increment in postgres).  So it's unlikely to be the same ID if you register in staging and then register in production.

Of course, adding a SHA hash to each message is much heavier than adding a global ID (a single int/long).  So depending on the typical size of a message, the SHA might add too much overhead...esp. in an IOT use-case.

**Comment by alesj** [comment]
public long toGlobalId(long offset, int partition) {
        return getBaseOffset() + (offset << 16) + partition;
    }

This is how we calculate globalId for Streams storage.
So if there is not more than 2^16 partitions, there cannot be dups.

**Comment by forsberg** [comment]
> I see - then I assume the data/schemas are separated from the apps such that the data layer has its own build? For example into a library that is then pulled into the apps when necessary? And then presumably the data is published to a per-app topic rather than a per-schema topic.
> 
> Because if the build is happening per-app then everything would hang together just fine using Apicurio Registry - the app build would be responsible for publishing the schema to the registry in the way that you describe. The only difference would be that multiple schemas (with the same content) would be published and therefore multiple global IDs would be created and packaged with the apps.

It can be a library, but it can also be part of an app build where the result is an artifact which is first installed in a test environment, then promoted to a production environment, the fine thing being that the schema ID doesn't have to change inside this artifact, it's fixed.

How information is then routed into Kafka is an implementation detail, it can be part of the endpoint the IoT device/app talks to, be it some gRPC endpoint or MQTT or whatever. The endpoint can use either a build number or the schema ID itself to route messages to the right place. I.e, you can run the same endpoint for production and staging, just route the messages to the right place.

>  I would need to read about the guarantees that the Confluent Schema Registry offers around global ID

The API reference is quite clear on this point (https://docs.confluent.io/current/schema-registry/develop/api.html#post--subjects-(string- subject)-versions):

_The returned identifier should be used to retrieve this schema from the schemas resource and is different from the schema’s version which is associated with the subject. If the same schema is registered under a different subject, the same identifier will be returned. However, the version of the schema may be different under different subjects._

I also did a little digging in the source code for canonical's schema registry, and from a quick glance, it looks like it's keeping an in-memory cache from the MD5 hash of the schema as sent by the client (not the canonical version) to the global schema ID, but don't take my word on this.

> 
> For global ID I'm worried about this moving of schema content from environment to environment and continuing to rely on that ID. In our system the ID is generated when the content is stored, based on the storage being used (e.g. an auto-increment in postgres). So it's unlikely to be the same ID if you register in staging and then register in production.

This is why, in #751, I'm asking for guidance on how to run multiple registries, with the production registry being the single point of truth for schema IDs, and registries reachable from staging environments being just a read-only replica.

In this way of working, a new schema, or a new schema version, is first registered under a staging subject. If the schema is considered correct, it is registered in the production subject, and the staging subject can be deleted - but the global schema ID stays the same in this operation.

Having the schema ID global even if registered under multiple subjects also makes communication within an organisation simpler, i.e. "What schema ID are you sending from the IoT device?" always has one answer, regardless of environment. 


> Of course, adding a SHA hash to each message is much heavier than adding a global ID (a single int/long). So depending on the typical size of a message, the SHA might add too much overhead...esp. in an IOT use-case.

Yes, if what we're sending is just a few temperature measurements (couple of integers or floats), the overhead is way too much.

So, to sum up, I would personally much rather have the same behaviour as in Confluent's registry, to the point where the apicurio behaviour might be a blocker from using it, I may have to go back to Confluent. 

If you decide on keeping the current behaviour, at the very least it needs to be clearly documented (right now, documentation claims apicurio registry is [compatible and drop-in](https://github.com/Apicurio/apicurio-registry/blob/bfcfe28d140b60526d3b49696a36253dc85ddf2d/docs/modules/getting-started/con-registry-rest-api.adoc#L19). Also, do consider how to handle this problem in #172, or users can get a nasty surprise when their global IDs change.

**Comment by EricWittmann** [comment]
First of all - thanks for taking the time to write all of this up.  Regardless of what we end up doing, this is very valuable information.

The pointer to the Confluent Schema Registry API docs is useful.  You are right, if we want to truly be "drop-in" then we need to consider this point, since it is clearly documented there.  That said, I don't think we'll be able to help the fact that we'll always have some inconsistencies between the implementations.  We need to decide which ones are important.  I think you're making a reasonable argument that this important.

Also it's interesting that they are using an MD5 hash of the original content (not canonicalized) to determine the globalId.  It seems like it would be better to use the canonical content rather than the original, but perhaps for most practical use-cases it doesn't matter.

What I was most concerned about with your explanation was the idea of migrating from e.g. staging to production.  I was assuming a different Schema Registry instance in those environments, rather than a shared one (and using a naming convention to separate staging content from production content).  My concern was how to get those globalIds to migrate from one instance to another consistently (hence the SHA suggestion).  But you're not actually doing that, so the point is moot.

I'll have to think about whether we can change the behavior without introducing any breaking changes.  If not, we might end up with a version 2.0.0.  Not the end of the world, just a consideration.

**Comment by katheris** [comment]
This is definitely an interesting discussion. I had also assumed that in general a different schema registry would be used in staging vs production, since my assumption was that application in staging wouldn't be able to access a production registry. I think the scenario that I hadn't considered before was having applications that only access the registry at build time. I guess that means that the applications do not have any mechanism to discover whether the schema they are using has been deprecated, disabled or removed entirely?

**Comment by EricWittmann** [comment]
That's a really good point, @katheris - I'm actually doing some work in the registry storage area right now, and I took some time to explore the consequences of this change (globalId associated with content, not a version).  Because the content becomes separated from the artifact version in a new many-to-one relationship, other meta-data (such as state) is no longer connected to the globalId, but instead to the artifactId+version.  So yeah - in general it means you can no longer lookup the version meta-data from just the globalId.  You can only get the content (and any content-specific meta-data we may choose to store with it such as "artifactType" maybe).

I can say already that we cannot make this change without introducing breaking API changes.  We definitely have a `/ids` endpoint in the Apicurio Registry API (which I sadly now realize is undocumented) that includes this:

`GET /ids/{globalId}/meta`

That wouldn't make sense anymore.  @katheris do you know if the IBM API uses the `globalId` concept at all?  I can't find any reference to it, but I could easily have missed it.

**Comment by EricWittmann** [comment]
We had a quick internal discussion about this today and realized that we don't necessarily have to implement this functionality for the Apicurio Registry API - we just need to implement it for the Confluent Compatibility API.  Thanks for that suggestion @jsenko 

I need to think through how we might be able to provide a separate "globalId" for the Confluent compatibility layer that behaves the way you are expecting and what implications that might have for other parts of the application and other use-cases.  For example, does this approach break the ability to mix and match Apicurio serdes classes with Confluent serdes classes?

So this is just an idea at this point, but if we can make it work then we'll have an option that will allow us to support the feature without doing a 2.0 release.

**Comment by EricWittmann** [comment]
Update:  this will be addressed in version 2.0.

@carlesarnal and @famartinrh - can you guys collaborate on making sure this is fixed?  I think there are two things to do here:

1) The ccompat API implementation needs to be updated to use `contentId` instead of `globalId` (I believe in all cases) ( @carlesarnal )
2) The Apicurio Registry Serdes classes needs a configuration option to read/write the `contentId` instead of the `globalId`

**Comment by EricWittmann** [comment]
@forsberg FYI :)
